File: app/src/main/java/com/kabulsignal/azanapp/ui/components/StateViews.kt
package com.kabulsignal.azanapp.ui.components

import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CloudOff
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.kabulsignal.azanapp.ui.theme.Radii
import com.kabulsignal.azanapp.ui.theme.Spacing

/** A single shimmering placeholder block. */
@Composable
fun SkeletonBlock(
    modifier: Modifier = Modifier,
    height: Dp = 20.dp
) {
    val transition = rememberInfiniteTransition(label = "skeleton")
    val alpha by transition.animateFloat(
        initialValue = 0.35f,
        targetValue = 0.8f,
        animationSpec = infiniteRepeatable(
            animation = tween(900),
            repeatMode = RepeatMode.Reverse
        ),
        label = "skeletonAlpha"
    )

    Box(
        modifier = modifier
            .height(height)
            .clip(Radii.sm)
            .alpha(alpha)
            .background(MaterialTheme.colorScheme.surfaceVariant)
    )
}

/**
 * Skeleton stand-in for the prayer list. A shaped placeholder beats a spinner here:
 * the layout does not jump when real data lands, so the first frame feels instant.
 */
@Composable
fun PrayerListSkeleton(modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(horizontal = Spacing.lg, vertical = Spacing.lg),
        verticalArrangement = Arrangement.spacedBy(Spacing.md)
    ) {
        SkeletonBlock(modifier = Modifier.fillMaxWidth(), height = 210.dp)
        Spacer(Modifier.height(Spacing.xs))
        repeat(5) {
            SkeletonBlock(modifier = Modifier.fillMaxWidth(), height = 68.dp)
        }
    }
}

/** Full-screen message with an icon and an optional action — used for errors and empty data. */
@Composable
fun MessageState(
    title: String,
    body: String? = null,
    icon: ImageVector = Icons.Default.CloudOff,
    actionLabel: String? = null,
    onAction: (() -> Unit)? = null,
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier.fillMaxSize(),
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier.padding(Spacing.xxl)
        ) {
            Surface(
                shape = Radii.pill,
                color = MaterialTheme.colorScheme.surfaceVariant,
                modifier = Modifier.size(72.dp)
            ) {
                Box(contentAlignment = Alignment.Center) {
                    Icon(
                        icon,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.size(32.dp)
                    )
                }
            }
            Spacer(Modifier.height(Spacing.lg))
            Text(
                text = title,
                style = MaterialTheme.typography.titleMedium,
                textAlign = TextAlign.Center,
                color = MaterialTheme.colorScheme.onSurface
            )
            if (body != null) {
                Spacer(Modifier.height(Spacing.sm))
                Text(
                    text = body,
                    style = MaterialTheme.typography.bodyMedium,
                    textAlign = TextAlign.Center,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            if (actionLabel != null && onAction != null) {
                Spacer(Modifier.height(Spacing.xl))
                Button(
                    onClick = onAction,
                    shape = Radii.pill,
                    modifier = Modifier.heightIn(min = Spacing.touchTarget)
                ) {
                    Icon(Icons.Default.Refresh, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.width(Spacing.sm))
                    Text(actionLabel, style = MaterialTheme.typography.labelLarge)
                }
            }
        }
    }
}

File: ios/Namazia/Components/StateViews.swift
import SwiftUI

/// A shimmering placeholder block, ported from `ui/components/StateViews.kt`.
struct SkeletonBlock: View {
    @Environment(\.colors) private var colors

    var height: CGFloat = 20
    var cornerRadius: CGFloat = Radii.sm

    @State private var bright = false

    var body: some View {
        RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
            .fill(colors.surfaceVariant)
            .frame(maxWidth: .infinity)
            .frame(height: height)
            .opacity(bright ? 0.8 : 0.35)
            .animation(
                .easeInOut(duration: 0.9).repeatForever(autoreverses: true),
                value: bright
            )
            .onAppear { bright = true }
    }
}

/// Stand-in for the prayer list while the first load runs.
///
/// A shaped placeholder rather than a spinner: the layout does not jump when the real
/// data lands, so the first frame already looks like the screen it is becoming.
struct PrayerListSkeleton: View {
    var body: some View {
        VStack(spacing: Spacing.md) {
            SkeletonBlock(height: 260, cornerRadius: Radii.xl)
            ForEach(0..<5, id: \.self) { _ in
                SkeletonBlock(height: 68, cornerRadius: Radii.md)
            }
            Spacer(minLength: 0)
        }
        .padding(Spacing.lg)
    }
}

/// Centred message with an icon and an optional action — errors and empty states.
struct MessageState: View {
    @Environment(\.colors) private var colors

    let title: String
    // Named `message`, not `body`: a stored property called `body` would collide with
    // the `View` requirement below.
    var message: String?
    var systemImage: String = "wifi.slash"
    var actionLabel: String?
    var action: (() -> Void)?

    var body: some View {
        VStack(spacing: Spacing.lg) {
            Image(systemName: systemImage)
                .font(.system(size: 30, weight: .regular))
                .foregroundStyle(colors.onSurfaceVariant)
                .frame(width: 72, height: 72)
                .background(colors.surfaceVariant, in: Circle())

            VStack(spacing: Spacing.sm) {
                Text(title)
                    .appText(AppType.titleMedium)
                    .foregroundStyle(colors.onSurface)

                if let message {
                    Text(message)
                        .appText(AppType.bodyMedium)
                        .foregroundStyle(colors.onSurfaceVariant)
                }
            }
            .multilineTextAlignment(.center)

            if let actionLabel, let action {
                Button(action: action) {
                    Label(actionLabel, systemImage: "arrow.clockwise")
                        .appText(AppType.labelLarge)
                        .padding(.horizontal, Spacing.lg)
                        .frame(minHeight: Spacing.touchTarget)
                }
                .buttonStyle(.borderedProminent)
                .tint(colors.primary)
                .clipShape(Capsule())
            }
        }
        .padding(Spacing.xxl)
        .frame(maxWidth: .infinity)
    }
}

/// Cached times are still shown when a refresh fails — but they are labelled rather
/// than passed off as live.
struct StaleDataNotice: View {
    @Environment(\.colors) private var colors

    let title: String

    var body: some View {
        Text("state.staleData".localized(title))
            .appText(AppType.bodySmall)
            .foregroundStyle(colors.onErrorContainer)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(Spacing.md)
            .background(
                colors.errorContainer.opacity(0.6),
                in: RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
            )
    }
}

extension View {
    /// The app's standard card: surface colour, rounded, comfortable padding.
    func cardSurface(
        _ colors: AppColors,
        cornerRadius: CGFloat = Radii.md,
        padding: CGFloat = Spacing.lg
    ) -> some View {
        self
            .padding(padding)
            .background(
                colors.surface,
                in: RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
            )
    }
}
